PR Fixing Issue I just Opened + What if Calculation, Among other things

Liam-Wirth opened the pull request:
(First PR, woohoo)



# Add what-if grades support & fix calculation issues

## Changes
1. Extended URL matching to support more Canvas instances:
   - Added `canvas.*.edu` domains
   - Added regex pattern for custom Canvas domains

2. Fixed grade calculation issues:
   - Added validation for assignment scores
   - Fixed NaN errors when dealing with zero scores
   - Improved handling of weighted categories
   - Added proper handling of what-if scores

3. Enhanced assignment parsing:
   - Broadened selector to include all assignments (`tr.student_assignment`)
   - Added fallback for various score formats
   - Improved handling of incomplete/ungraded assignments
   - Better null-safety with optional chaining

4. Added what-if grade UI:
   - Added recalculation button
   - Shows original grade for comparison
   - Updates in real-time when what-if grades are entered
   - Maintains grade history



## Key changes/relevant description regarding changes made with the NaN fix included in Issue #5 that I opened.

## Key Changes with What-If Calculation
Was unable to implement a dom listener, but updated the selection code to select a wider range of assignments, and then when we actually need to do calculation in `calculate()` filter out those that would be "invalid" (I.E null/ungraded etc)

Before:
![image](https://github.com/user-attachments/assets/1f9cc6a3-afe0-4ed7-8d4f-f2a065b1c2b5)


After:
 1. Before the user enters any what if grade values
![image](https://github.com/user-attachments/assets/38ee9a62-f095-494a-be8e-4c368315d3dc)
 2. After editing/changing a value and clicking the button:
![image](https://github.com/user-attachments/assets/b41c2103-c22c-4058-b466-21c84430c5e7)



Again this is my first time creating a PR on anything public so I'm not sure what else to include. I hope this is helpful enough

Liam-Wirth committed:
(fix + feat) Fixed NaN error on my end by javascript being javascript (0 is apparently a Falsy value, so in the grade weight calculation, when you fall back to \unassigned if the weight of the assignment was also 0, it would be considered unassigned as well, and cause grades to be multiplied by a non numerical value. Also added functionality to recalculate current grade after calculation to support canvas what-if grades

Liam-Wirth committed:
(feat) Got what if grades working properly by making the jquery select search broader in the getassignments function. This includes assignments whose 'earned' value is null, to mitigate this the calculate() function filters out all assignments that would be deemed 'invalid'

(feat) cached original grade so you can see how it differs from the calculated what if grade

Liam-Wirth committed:
(feat) added regex for more unique canvas URLs, my university has a different way of hosting canvas

uncenter commented on index.js:
```diff
@@ -3,12 +3,15 @@
+// @author      uncenter, Liam Wirth
 // @license     MIT
+// @downloadURL https://update.greasyfork.org/scripts/479317/Canvas%20Grade%20Calculator.user.js
+// @updateURL https://update.greasyfork.org/scripts/479317/Canvas%20Grade%20Calculator.meta.js
```
This shouldn't be needed? I think the userstyles manager handles tracking the update URL automatically, at least it does with Violentmonkey iirc.

uncenter commented on index.js:
```diff
@@ -3,12 +3,15 @@
 // @description Calculate grade totals for Canvas courses that have it disabled.
 // @namespace   https://github.com/uncenter/canvas-grade-calculator
 // @match       https://*.instructure.com/courses/*/grades
+// @match       https://canvas.*.edu/courses/*/grades
+// @include     /^https:\/\/canvas\..*\.edu\/courses\/.*\/grades/
```
Do we need both?

uncenter reviewed (changes requested):
This repository uses [Prettier](https://prettier.io/) for code formatting. You can run `pnpm install` and `pnpm format` to fix this.

Liam-Wirth committed:
(fix) Formatting + Removing redundant match pattern

Removing auto-added lines

Liam-Wirth commented:
Whoops! entirely forgot to run prettier/format the code. My bad this likely happened cause I was doing all of my editing within the Violentmonkey script editor and forgot  to actually format anything within my regular text editor before pushing.

As for the issue you mentioned regarding the two URLs, I removed the simpler one and kept the regex, as apparently the `@match` operator only focuses on the scheme, host, and path of a URL. Not supporting full regex/wildcards in the middle of domain components. So a line like 
```js
// @match canvas.*.edu
```
would not work, as the wildcard operator is being used within the top level domain, which is unsupported for some reason.
After perusing docs I found an older `@include` directive which works similarly to `@match` but allows for full regex,
so I wrote up the following:
```js
// @include /^https:\/\/canvas\..*\.edu\/courses\/.*\/grades/
```

I get a warning within the editor saying it can be a bit slower to use, but they say that the difference is marginal, and in my experience I didn't notice any noticeable load time changes between using the regex, and a hardcoded TLD for my university. 

In my updated push I've gone ahead and removed the non-functioning/broken match statement and opted to keep the @include, if you decide to remove it it's no worries cause I can just edit the script down the line on my end.

As for the second comment regarding the 
```js
// @downloadURL https://update.greasyfork.org/scripts/479317/Canvas%20Grade%20Calculator.user.js
// @updateURL https://update.greasyfork.org/scripts/479317/Canvas%20Grade%20Calculator.meta.js
```
lines I don't remember adding them into the script, and apparently greasyfork [adds those lines automatically](https://greasyfork.org/en/discussions/greasyfork/214165-downloadurl-and-uploadurl-shown-in-tampermonkey-editor-not-on-greasyfork) downstream when a script gets updated/installed.

uncenter committed:
chore: delete package-lock.json

uncenter commented on index.js:
```diff
@@ -85,88 +85,120 @@ function getWeights() {
-				)
-				.getAttribute('style') === 'visibility: hidden;';
+		// Determine if the assignment counts toward the final grade.
+		let countsTowardFinalGrade = true;
+		const ctfgEl = row.querySelector(
+			'[aria-label="This assignment does not count toward the final grade."]'
+		);
+		if (ctfgEl) {
+			countsTowardFinalGrade =
+				ctfgEl.getAttribute('style') === 'visibility: hidden;';
+		}
```
```suggestion
		let countsTowardFinalGrade = row.querySelector(
			'[aria-label="This assignment does not count toward the final grade."]'
		)?.getAttribute('style') === 'visibility: hidden;';
```

uncenter commented on index.js:
```diff
@@ -85,88 +85,120 @@ function getWeights() {
 
 function getAssignments() {
 	const assignments = [];
+	// Use a broader selector to include all assignment rows,
+	// not just those already marked as graded.
```
This comment uses a tone/tense of a PR comment, "broader" doesn't make sense after being merged since this has no reference. Use absolute wording like "Select all assignments (regardless of grade status) from assignments table."

uncenter commented on index.js:
```diff
@@ -85,88 +85,120 @@ function getWeights() {
 		weights[group] = weight;
 	}
-
+	console.log(weights);
```
```suggestion
```
Looks like a debug line.

uncenter commented on index.js:
```diff
@@ -85,88 +85,120 @@ function getWeights() {
 	const assignments = [];
+	// Use a broader selector to include all assignment rows,
+	// not just those already marked as graded.
+	const assignmentRows = document.querySelectorAll(
+		'#grades_summary tr.student_assignment'
+	);
```
```suggestion
```

uncenter commented on index.js:
```diff
@@ -85,88 +85,120 @@ function getWeights() {
-			earned = Number.parseFloat(
-				assignment.querySelector('.original_points').textContent.trim()
-			);
+	for (const row of assignmentRows) {
```
```suggestion
	for (const assignmentEl of document.querySelectorAll(
		'#grades_summary tr.student_assignment'
	)) {
```

uncenter commented on index.js:
```diff
@@ -85,88 +85,120 @@ function getWeights() {
-				assignment.querySelector('.original_points').textContent.trim()
-			);
+	for (const row of assignmentRows) {
+		let earned = 0,
+			available = 0,
+			title = '',
+			group = '';
```
I don't see the point in setting initial values here.

uncenter commented on index.js:
```diff
@@ -85,88 +85,120 @@ function getWeights() {
+			group = '';
+
+		// Get assignment title and group from the header cell.
+		const titleEl = row.querySelector('th.title');
```
```suggestion
		const titleEl = assignmentEl.querySelector('th.title');
```

uncenter commented on index.js:
```diff
@@ -85,88 +85,120 @@ function getWeights() {
+		// Get assignment title and group from the header cell.
+		const titleEl = row.querySelector('th.title');
+		if (!titleEl) continue;
+		title = titleEl.querySelector('a')?.textContent.trim() || '';
```
I don't see the point in setting a backup value here. I would prefer an error earlier in the chain if this is invalid/incorrect.

uncenter commented on index.js:
```diff
@@ -85,88 +85,120 @@ function getWeights() {
+		title = titleEl.querySelector('a')?.textContent.trim() || '';
+		group = titleEl.querySelector('div.context')?.textContent.trim() || '';
+
+		// Parse due and submitted dates.
+		const dueText = row.querySelector('.due')?.textContent.trim() || '';
+		const submittedText =
+			row.querySelector('.submitted')?.textContent.trim() || '';
+		const due = parseCanvasDate(dueText);
+		const submitted = parseCanvasDate(submittedText);
```
It was just fine as before. This change is unnecessary.

Liam-Wirth replied:
Resolved, reverted to previous version.

uncenter commented on index.js:
```diff
@@ -85,88 +85,120 @@ function getWeights() {
+
+		// Get the grade container.
+		const gradeEl = row.querySelector('td.assignment_score span.grade');
+		if (!gradeEl) continue;
```
I'd rather throw so we know something happened rather than silently ignoring it.

uncenter commented on index.js:
```diff
@@ -85,88 +85,120 @@ function getWeights() {
+
+		// Get assignment title and group from the header cell.
+		const titleEl = row.querySelector('th.title');
+		if (!titleEl) continue;
```
```suggestion
```
I'd rather throw so we know something happened rather than silently ignoring it.

uncenter commented on index.js:
```diff
@@ -85,88 +85,120 @@ function getWeights() {
+		// If the grade element shows a graded icon, use the original points.
+		if (gradeEl.querySelector('.graded_icon')) {
+			const op = row.querySelector('.original_points');
+			earned = op ? Number.parseFloat(op.textContent.trim()) : 0;
```
In what case is setting a backup value necessary?

Liam-Wirth replied:
Helps out with these edge case:
    The .what_if_score element exists but contains an empty string or something unparseable.
    original_points is missing or malformed.
    The extracted score text is NaN, which would otherwise propagate errors in later calculations.

Mostly just as a failsafe. By setting anything we get to zero it would get filtered out/not contribute to the overall grade calculation later down the line..

Liam-Wirth replied:
you are right though, in this context removing the failsafe did nothing when testing. It was likely redundant code.

uncenter commented on index.js:
```diff
@@ -85,88 +85,120 @@ function getWeights() {
+
+		// If the grade element shows a graded icon, use the original points.
+		if (gradeEl.querySelector('.graded_icon')) {
+			const op = row.querySelector('.original_points');
```
I don't think extracting this into a variable is necessary. Since this is my codebase I'd prefer if we stick to my code style :)

Liam-Wirth replied:
Fixed

uncenter commented on index.js:
```diff
@@ -85,88 +85,120 @@ function getWeights() {
+			// For ungraded or "what-if" assignments, try to extract a score text.
+			const textNode = [...gradeEl.childNodes].find(
+				(node) =>
+					node.nodeType === Node.TEXT_NODE && node.textContent.trim() !== ''
+			);
 
-				available = Number.parseFloat(
-					grades.nextElementSibling.textContent.replace('/', '').trim()
-				);
+			if (textNode) {
+				const scoreText = textNode.textContent.trim();
```
This was just fine as before with `?.textContent.trim();` instead of a separate if block.

uncenter commented on index.js:
```diff
@@ -85,88 +85,120 @@ function getWeights() {
-					name: name.trim(),
-					date,
-				});
+		const commentContainer =
+			row.nextElementSibling &&
+			row.nextElementSibling.nextElementSibling &&
+			row.nextElementSibling.nextElementSibling.nextElementSibling;
```
These checks appear unnecessary. Did you run into errors before?

uncenter commented on index.js:
```diff
@@ -181,7 +213,7 @@ function getAssignments() {
 		});
 	}
-
+	console.log(assignments);
```
```suggestion
```

uncenter commented on index.js:
```diff
@@ -85,88 +85,120 @@ function getWeights() {
+			row.nextElementSibling.nextElementSibling &&
+			row.nextElementSibling.nextElementSibling.nextElementSibling;
+		if (commentContainer) {
+			const commentTable = commentContainer.querySelector('td > table > tbody');
+			if (commentTable) {
+				for (const commentRow of commentTable.querySelectorAll('tr')) {
+					const [commentTd, detailsTd] = commentRow.querySelectorAll('td');
+					const text =
+						commentTd.querySelector('span')?.textContent.trim() || '';
+					let [name, ...dateParts] = detailsTd.textContent.trim().split(',');
+					const date = parseCanvasDate(dateParts.join(',').trim());
+					comments.push({
+						text,
+						name: name.trim(),
+						date,
+					});
+				}
```
Again please keep in mind my style of writing code. This appears to have little functional difference - I appreciate functional improvements, I'm less inclined for the style changes.

uncenter commented on index.js:
```diff
@@ -194,9 +226,17 @@ function calculate() {
 		return;
 	}
 
+	// Filter out assignments where earned or available is not a finite number.
+	const validAssignments = assignments.filter(
+		(a) => Number.isFinite(a.earned) && Number.isFinite(a.available)
+	);
```
Finite number as opposed to...?

Liam-Wirth replied:
This calls back to the NaN issue I was having before. Sometimes a professor/teacher in canvas will opt to fill out all remaining assignments in the semester. In canvas lets say we have some assignment foo, that would eventually be graded out of 100, if no grade was entered/submitted for the assignment, the current grade would be either NaN or unassigned (can't remember which). This helps clean that up, the query selector I set up before (to enable what-if calculation) grabs all assignments listed, regardless of whether or not they have an actual grade assigned to them

uncenter replied:
We shouldn't need to check if numbers are "finite" though. If the current grade is valid at parsing we should set it to undefined, and then only filter for not-undefined values for valid assignments. Also for the grade export are we exporting non "valid" assignments now? I think that would be an improvement (not only exporting the graded assignments).

uncenter commented on index.js:
```diff
@@ -194,9 +226,17 @@ function calculate() {
+		(a) => Number.isFinite(a.earned) && Number.isFinite(a.available)
+	);
+
+	console.log('These assignments are valid');
+	console.log(validAssignments);
```
I'd like to keep the console clear as much as possible. If you want to we can set up a `debug` function that can be enabled via some flag in a separate PR.

uncenter commented on index.js:
```diff
@@ -194,9 +226,17 @@ function calculate() {
 	/* eslint-disable unicorn/prevent-abbreviations, unicorn/no-array-reduce */
-	// Convert assignments array into an object of type { [category]: { totalEarned: number, totalAvailable: number }].
-	const totalsPerGroup = assignments
+	// Build totals per group from only valid assignments.
```
I'd like to keep the type information since `reduce` can be confusing for the casual reader. You can use JSDoc if you want.

uncenter commented on index.js:
```diff
@@ -206,16 +246,18 @@ function calculate() {
 	/* eslint-enable */
 
-	// Convert available out of total for each group into percentage values.
+	// Calculate each group's percentage.
```
No reason to change comment.

uncenter commented on index.js:
```diff
@@ -206,16 +246,18 @@ function calculate() {
 	let grade = 0;
 	if (Object.entries(weights).length === 0) {
-		// No weights, so we can just take total earned and available combined from all groups and get the percentage.
+		// No weights: combine all valid assignments.
```
This version of the comment is more vague, not sure why we need to change it.

uncenter commented:
For pull requests across all kinds of projects it is generally preferred to split up many little changes into separate PRs to make reviewing easier and more efficient. Some of these changes are great as is and I would happily merge quickly, others need more time - helps everyone to split it up.

Liam-Wirth committed:
Update index.js

Use Optional-Chaining for a cleaner/more concise query selector

Co-authored-by: uncenter <uncenter@uncenter.dev>

Liam-Wirth committed:
Update index.js

Removal of debug lines

Co-authored-by: uncenter <uncenter@uncenter.dev>

Liam-Wirth committed:
Update index.js

Co-authored-by: uncenter <uncenter@uncenter.dev>

Liam-Wirth committed:
Update index.js

Co-authored-by: uncenter <uncenter@uncenter.dev>
